### PR TITLE
Unwind cleanup

### PR DESCRIPTION
**Pull request by jforissier**

### CODE REVIEW

File: lib/libunw/unwind_arm32.c [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+{
+	trace_printf_helper_raw(TRACE_ERROR, true, "Call stack:");
+	do {
+		trace_printf_helper_raw(TRACE_ERROR, true, " 0x%08" PRIx32,
+					state->registers[PC]);
+	} while (unwind_stack_arm32(state, stack, stack_size));
+}
+
+#if defined(__KERNEL__) && defined(ARM32)
+bool find_exidx(vaddr_t addr __unused, vaddr_t *idx_start, vaddr_t *idx_end)
```
Comment (jenswi-linaro) [inline lib/libunw/unwind_arm32.c:425]: Wouldn't it make more sense to keep this outside libunw?
This depends on internals of Core. For ldelf the counterpart is actually in ldelf.
Same goes for for the other functions below, they seem more suited to keep in Core.

**Review by etienne-lms — COMMENTED** [review]
Some minor style comments for commit "Add unwind library (libunw)", otherwise looks all good to me. Succesfully tested on an armv7 platform with TAs rebuilt and not rebuilt against these changes.

File: lib/libunw/unwind_arm32.c [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+
+	return addr;
+err:
+	EMSG("Out of memory");
+	return NULL;
+}
+
+void print_kernel_stack(void)
+{
+	struct unwind_state_arm32 state = {};
```
[the remark this answers is not in the document]
Comment (jforissier) [reply lib/libunw/unwind_arm32.c:486]: ok

File: lib/libunw/unwind_arm64.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+#if defined(__KERNEL__) && defined(ARM64)
+vaddr_t *unw_get_kernel_stack(void)
+{
+	size_t n = 0;
+	size_t size = 0;
+	vaddr_t *tmp = NULL;
+	vaddr_t *addr = NULL;
+	struct unwind_state_arm64 state = { 0 };
```
[the remark this answers is not in the document]
Comment (jforissier) [reply lib/libunw/unwind_arm64.c]: ok

File: lib/libunw/unwind_arm64.c [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	return addr;
+err:
+	EMSG("Out of memory");
+	free(addr);
+	return NULL;
+}
+
+void print_kernel_stack(void)
+{
+	struct unwind_state_arm64 state = {};
```
[the remark this answers is not in the document]
Comment (jforissier) [reply lib/libunw/unwind_arm64.c:137]: ok

File: lib/libunw/unwind_private.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+#include <stddef.h>
+#include <util.h>
+
+static inline void *unw_grow(void *p, size_t *cur_size, size_t new_size)
+{
+	if (*cur_size >= new_size)
+		return p;
+
+	size_t rounded_size = ROUNDUP(new_size, 16 * sizeof(vaddr_t));
+	void *tmp = realloc(p, rounded_size);
```
[the remark this answers is not in the document]
Comment (jforissier) [reply lib/libunw/unwind_private.h:15]: ok

**Comment by jforissier** [comment]
Update. Let me know if you prefer that I squash the fixups (code moving back and forth is not easy to review...).

**Comment by jenswi-linaro** [comment]
Please squash.

**Comment by jforissier** [comment]
Update

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: lib/libunw/unwind_arm32.c [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+ * with the largest address that does not exceed addr.
+ */
+static struct unwind_idx *find_index(uint32_t addr)
+{
+	vaddr_t idx_start, idx_end;
+	unsigned int min, mid, max;
+	struct unwind_idx *start;
+	struct unwind_idx *item;
+	int32_t prel31_addr;
+	vaddr_t func_addr;
```
Comment (etienne-lms) [inline lib/libunw/unwind_arm32.c]: sorry, my comment here was discarded. Would you mind adding default initialization values for these local variables. Here and in the functions below. Unless you think it's better to stick to the original implementation...

File: lib/libunw/unwind_arm32.c [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+
+				/* If we have updated SP kep its value */
+				if (reg == SP)
+					update_vsp = 0;
+			}
+		}
+
+	} else if ((insn & INSN_STD_MASK) == INSN_VSP_REG &&
+	    ((insn & INSN_STD_DATA_MASK) != 13) &&
+	    ((insn & INSN_STD_DATA_MASK) != 15)) {
```
Comment (etienne-lms) [inline lib/libunw/unwind_arm32.c]: indentation

**Review by etienne-lms — COMMENTED** [review]
Minor coding style comments for commit "Add unwind library (libunw)", to address or not...

Typo in commit log for "core: use libunw":
```patch
-Reduce core/arch/arm/kernel/unwind_arm32,64}.c and use common code from
+Reduce core/arch/arm/kernel/unwind_arm{32,64}.c and use common code from
```

That said, `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`.

File: lib/libunw/unwind_arm32.c (hunk 6 above)
Comment (jforissier) [reply lib/libunw/unwind_arm32.c]: ok

File: lib/libunw/unwind_arm32.c (hunk 7 above)
Comment (jforissier) [reply lib/libunw/unwind_arm32.c]: right, will fix

**Comment by jforissier** [comment]
Latest comments addressed and tags applied, thanks!

**Pull request merged**